python/ray/data/_internal/execution/operators/actor_pool_map_operator.py

```diff
@@ -529,31 +590,43 @@ def penalty_key(actor):
             We prioritize valid actors, those with argument locality, and those that
             are not busy, in that order.
             """
-            busyness = self._num_tasks_in_flight[actor]
-            invalid = busyness >= self._max_tasks_in_flight
-            requires_remote_fetch = self._actor_locations[actor] != preferred_loc
+            busyness = self._running_actors[actor]._num_tasks_in_flight
+            is_restarting = self._running_actors[actor]._is_restarting
+            invalid = busyness >= self._max_tasks_in_flight or is_restarting
+            requires_remote_fetch = (
+                self._running_actors[actor]._actor_location != preferred_loc
+            )
             return invalid, requires_remote_fetch, busyness
 
-        actor = min(self._num_tasks_in_flight.keys(), key=penalty_key)
-        if self._num_tasks_in_flight[actor] >= self._max_tasks_in_flight:
-            # All actors are at capacity.
+        actor = min(self._running_actors.keys(), key=penalty_key)
+        if (
+            self._running_actors[actor]._num_tasks_in_flight
+            >= self._max_tasks_in_flight
+            or self._running_actors[actor]._is_restarting
+        ):
+            # All actors are at capacity or restarting.
             return None
 
         if locality_hint:
-            if self._actor_locations[actor] == preferred_loc:
+            if self._running_actors[actor]._actor_location == preferred_loc:
                 self._locality_hits += 1
             else:
                 self._locality_misses += 1
-        self._num_tasks_in_flight[actor] += 1
+        self._running_actors[actor]._num_tasks_in_flight += 1
         return actor
 
     def return_actor(self, actor: ray.actor.ActorHandle):
         """Returns the provided actor to the pool."""
-        assert actor in self._num_tasks_in_flight
-        assert self._num_tasks_in_flight[actor] > 0
+        assert actor in self._running_actors
+        assert self._running_actors[actor]._num_tasks_in_flight > 0
 
-        self._num_tasks_in_flight[actor] -= 1
-        if self._should_kill_idle_actors and self._num_tasks_in_flight[actor] == 0:
+        # Mark restarting as false, now that the actor in running
+        self._running_actors[actor]._is_restarting = False
```

Comment:
after a second thought, I think it'd be slightly more clear to remove this and let the next update_resource_usage to handle the state transition